### DISCUSSION TITLE
Heterogenous Roughness

### DISCUSSION THREAD

**Discussion by hgopalan** [Exawind/amr-wind#1116] [Ideas]
Currently we can prescribe only a single roughness length for entire simulations. However, for many situations, this is not idea and has to be modified. This would remove a modification of how the friction velocity and shear stresses are committed. I have an initial version of this change which goes through the following steps: 

1. Read roughness length from a file. 
2. Compute the friction velocity at every point by modifying the ABLWallFunction.cpp routines. 

I wanted to check to see if others have explored heterogenous roughness and have a different way of implementing the changes.

**Comment by lawrenceccheung** [comment]
Hi @hgopalan, 

This is a great idea and will be very useful in the future.  It will also be great to generalize this to the surface heat flux as well.  I'm curious, how does the variable roughness length get consistently applied across all refinement levels in the wall model formulation?

Lawrence

**Comment by hgopalan** [comment]
I have implemented a method for the heterogenous terrain roughness within the terrain model: https://github.com/hgopalan/amr-wind/blob/hgopalan-terrain/amr-wind/physics/TerrainDrag.cpp

For a general implementation we need three steps: 

1. Create a new physics model to read the roughness length and heat flux and map it to a field variable similar to terrain height. 2D field variable would have been ideal but after discussing with Micheal, there seems to be no easy way to do it in AMR-Wind. 
2. Read the roughness and heat flux value within each level 
3. Replace the calls in https://github.com/hgopalan/amr-wind/blob/hgopalan-terrain/amr-wind/wind_energy/ABLWallFunction.cpp to directly compute u* and theta* within the code instead of calling calc_vel_x, calc_vely and calc_theta.

**Comment by hgopalan** [comment]
Capability is now available.
